internal/consensus/pbts_test.go

```diff
@@ -1,56 +1,299 @@
+		res := heightResult{}
+		for {
+			voteMsg := <-voteSub.Out()
+			ts := time.Now()
+			vote := voteMsg.Data().(types.EventDataVote)
+			if !bytes.Equal(address, vote.Vote.ValidatorAddress) {
+				continue
+			}
+			voteEvent, _ := voteMsg.Data().(types.EventDataVote)
+			if voteEvent.Vote.Type != tmproto.PrevoteType {
+				continue
+			}
+			res.prevoteIssuedAt = ts
+			res.prevote = voteEvent.Vote
+			break
+		}
+		err := eb.UnsubscribeAll(context.Background(), "voteSubscriber")
+		assert.NoError(t, err)
+		resultCh <- res
+		close(resultCh)
+	}()
+	return resultCh
+}
+
+func (p *pbtsTestHarness) run(t *testing.T) resultSet {
+	p.genesisHeight(t)
+	r2 := p.height2(t)
+	return resultSet{
+		height2: r2,
+	}
+}
+
+type resultSet struct {
+	height2 heightResult
+}
+
+type heightResult struct {
+	prevote         *types.Vote
+	prevoteIssuedAt time.Time
+}
+
+// TestReceiveProposalWaitsForPreviousBlockTime tests that a validator receiving
+// a proposal waits until the previous block time passes before issuing a prevote.
+// The test delivers the block to the validator after the configured `timeout-propose`,
+// but before the proposer-based timestamp bound on block delivery and checks that
+// the consensus algorithm correctly waits for the new block to be delivered
+// and issues a prevote for it.
+func TestReceiveProposalWaitsForPreviousBlockTime(t *testing.T) {
+	initialTime := time.Now().Add(50 * time.Millisecond)
+	cfg := pbtsTestConfiguration{
+		timestampParams: types.TimestampParams{
+			Precision: 100 * time.Millisecond,
+			MsgDelay:  500 * time.Millisecond,
+		},
+		timeoutPropose:             50 * time.Millisecond,
+		genesisTime:                initialTime,
+		height2ProposalDeliverTime: initialTime.Add(450 * time.Millisecond),
+		height2ProposedBlockTime:   initialTime.Add(350 * time.Millisecond),
+	}
+
+	pbtsTest := newPBTSTestHarness(t, cfg)
+	results := pbtsTest.run(t)
+
+	// Check that the validator waited until after the proposer-based timestamp
+	// waitinTime bound.
+	assert.True(t, results.height2.prevoteIssuedAt.After(cfg.height2ProposalDeliverTime))
+	maxWaitingTime := cfg.genesisTime.Add(cfg.timestampParams.Precision).Add(cfg.timestampParams.MsgDelay)
+	assert.True(t, results.height2.prevoteIssuedAt.Before(maxWaitingTime))
+
+	// Check that the validator did not prevote for nil.
+	assert.NotNil(t, results.height2.prevote.BlockID.Hash)
+
+}
+
+// TestReceiveProposalTimesOutOnSlowDelivery tests that a validator receiving
+// a proposal times out and prevotes nil if the block is not delivered by the
+// within the proposer-based timestamp algorithm's waitingTime bound.
+// The test delivers the block to the validator after the previous block's time
+// and after the proposer-based timestamp bound on block delivery.
+// The test then checks that the validator correctly waited for the new block
+// and prevoted nil after timing out.
+func TestReceiveProposalTimesOutOnSlowDelivery(t *testing.T) {
+	initialTime := time.Now()
```

Comment:
Is this safely miscible with the fake used internally by the test harness?

Reply:
The fake clock is set using the values provided in the configuration fields. `initialTime` is just captured here to make it so that all of the timestamps set on the test configuration would be offset from the same value instead of all being offset from  whenever the `time.Now` call was evaluated. Does this answer your question?

Reply:
It does, esp. combined with our discussion offline.